### DISCUSSION TITLE
Allow mailer to change default sender name

### DISCUSSION THREAD

**Discussion by mattiaz9** [directus/directus#11042] [Feature Requests]
Right now the name of the sender is the project name which is constraining. If someone has multi-domain project or simply wants to send mails for different scopes (eg: marketing campaign, direct contact ...) he can't change the name.

That would be a simple change of the `mail/index.ts` file from:
```typescript
const from = `${defaultTemplateData.projectName} <${options.from || (env.EMAIL_FROM as string)}>`;
```
to:
```typescript
const from = `${options.fromName || defaultTemplateData.projectName} <${options.from || (env.EMAIL_FROM as string)}>`;
```

**Comment by unleashit** [comment]
Just came across this after trying to figure out why "Directus" was always being prepended to `EMAIL_FROM`. It was because I was temporarily calling my project Directus 🤣 .

I agree with the poster this could be a nice feature. But I think it would at least be nice to document under https://docs.directus.io/self-hosted/config-options.html#email that the convention of designating the sender name and address together in "from email" is being overridden. For example, when I tried to set `EMAIL_FROM` like:

```text
EMAIL_FROM='Company <no-reply@company.com>', this resulted in the "From" line:

From: "Directus >" <no-reply@company.com> (from the email header).
```

Had to scratch my head for a bit until finding this.

**Reply by unleashit** [reply]
Just went to try and make a PR with an edit the docs, but couldn't find them in the repo. If there's a different public repo somewhere I'd be happy to do it.

**Reply by Lukas-Sachse** [reply]
They decided to make it private some time ago. @phazonoverload can help you, he is working on the docs.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by Pachat** [comment]
@mattiaz9  In fact, for some reason,
```
const from = `${options.fromName || defaultTemplateData.projectName} <${options.from || (env.EMAIL_FROM as string)}>`;
```

has to be

```
const from = `${options.fromName || defaultTemplateData.projectName} <${options.from || (env.EMAIL_FROM as string)}`;
```

(without the `>`). 
Because for some reason a `>` is added later in the Directus code.

**Comment by christianrr** [comment]
We would also need a possibility to override the from name in MailService's `send` method .

I would suggest to refine the code as follows. If only a plain email address is passed, it will be combined with the project name (current behavior). Otherwise, it will simply use the formatted name that is passed.

This would allow for using both plain and formatted addresses in both the `EMAIL_FROM` env variable as well as in the `options.from` parameter. Moreover, it would support passing Nodemailer's Address object in `options.from`.

I think this would provide most flexibility for future use-cases and is also fully downwards compatible.

```
let from = env['EMAIL_FROM'] as string;

// Nodemailer's EmailOptions for from support either plain addresses, formatted names or Address objects
if (options.from) {
	if (typeof options.from === 'string') {
		from = options.from;
	} else if (typeof options.from === 'object' && options.from.name && options.from.address)
		from = `${options.from.name} <${options.from.address}>`;
}

// From does not include formatted name, so construct one
if (from && !from.includes('<') && !from.includes('>')) from = `${defaultTemplateData.projectName} <${from}>`;
```

I would have a feature branch ready and could start a pull request. However, there is currently no issue open.

**Reply by christianrr** [reply]
@rijkvanzanten What do you think, can I start a Pull Request?

**Reply by glowkeeper** [reply]
Currently, I've set up a password-reset.liquid template, and it works nicely, except the from address ends up being **Directus \<my@emailaddress.com\>**, which looks totally unprofessional - much better would be **My Name \<my@emailaddress.com\>**. But short of writing my own password reset extensions (disclaimer - I did that on a past Directus project, but I'd prefer to avoid that with my current project), I can't see how to change the name. Is that what this thread is about? If so, can it be actioned, please?

**Reply by glowkeeper** [reply]
Aha! So I reread the original post, got a clue from "It was because I was temporarily calling my project Directus 🤣", found that setting, changed it, and now I have **My Name \<my@emailaddress.com\>**, which is much better - huzzah!

My problem now is that, much like the original postee, I'm using Directus as a backend proxy for some API services across multiple projects, and I want to send reset emails from those different projects - it would be great if I were able to assign different project names to those reset emails...

**Comment by christianrr** [comment]
Can we continue discussion on this? It is still not possible to override the sender name in `MailerService`'s `send` method, as discussed in https://github.com/directus/directus/issues/20053. I can still see a need for this within the community (including us).

@br41nslug mentioned that the suggested change would be a breaking change for all that use `from` to override the email address.
My suggestion from above would take this into account and still allow for various different usage scenarios.

The following use cases would be supported:

by using `from` in `send(...)` of `MailService`:
- to set the email address only, e.g. `email@domain.com` becomes `Directus Project Name <email@domain.com>`
- to use a formatted name, e.g. `My Sender Name <email@domain.com>`
- to use a Nodemailer Address Object, e.g. `{ name: 'My Sender Name', address: 'email@domain.com' }` becomes `My Sender Name <email@domain.com>`
https://nodemailer.com/message/addresses/

by using environment variable `EMAIL_FROM`
- to set email address only, e.g. `email@domain.com` becomes `Directus Project Name <email@domain.com>`
- to use a formatted name, e.g. `My Sender Name <email@domain.com>`

I don't see another use case that would not be supported.

Fix would be ready for PR here:
https://github.com/christianrr/directus/tree/fix-formatted-email-addresses

**Reply by rijkvanzanten** [reply]
The reason I've been hesitant with introducing more flexibility into the email sending headers (incl. from name) is that it very quickly opens up security problems and domain reputation problems. A lot of Directus instances run with users that may or may can be trusted, so having an email from header that maliciously can be set to something different than the sysadmin intended can very easily cause big issues..

**Reply by jekuer** [reply]
@rijkvanzanten , what do you think about at least extending the mailservice, so we can write extensions to use more nodemailer options?
At the moment, I have an extension in place, which basically bridges the mailservice with a mailservice-copy that I extended. Would be nice to have the "native" mailservice support more options.
By only extending it on the code level, I do not see big security risks, as one would still need to connect to that on a rather deep level (custom extension), which is nothing a regular user would do (or, when thinking of cloud users, even can do).

**Reply by jekuer** [reply]
This would also resolve #19989

**Reply by rijkvanzanten** [reply]
The concerns from my post above would still persist, but I'm getting to a point being fine with just saying it's a use-at-your-own-risk type of thing.

> as one would still need to connect to that on a rather deep level (custom extension)

I'm just concerned about folk that download/use a third party built extension that then causes security issues. Can't prevent all bad uses though, and I don't want to hurt the core DX too much. Happy to take a PR to open up the mail service with some additional options for use in extensions 👍🏻

**Reply by jekuer** [reply]
@rijkvanzanten , sorry for getting back to this so late.

I went a little **deeper**. 
The only things that are not already changeable are the from address as well as the transport options. Both would be an issue if one would want to create an extension to send emails via different systems (e.g. having the default for Directus itself, while using the same Mailer for sending email via another address and host - imagine something like sending newsletters from a Flow).

My **suggestion** would be to have an override option at the MailService. Basically an option, where an extension author can override the defaults (that come via env vars) temporarily. Would also only apply to explicit settings, not the ones that are pulled rather dynamically (like the ses vars).

**I already prepared the changes and could start a PR - however, feels like there are only discussions about it - no issue 😉 - how do you want me to proceed?**

**Reply by rijkvanzanten** [reply]
@jekuer The main problem we have to prevent here is spam filters denylisting the email server you're using. Sending emails from multiple different from-addresses from the same email server is a surefire way to get in trouble with spam protections. Once we allow all those transport options to be changed, we have to require a custom SMTP or other email server to make sure you don't tank the reputation on the main transactional one 🤔

**Reply by jekuer** [reply]
@rijkvanzanten , that's true, but isn't that something the "operator" needs to consider when configuring the system?

In my implementation, one can override the FROM address and also the SMTP settings - basically because of what you wrote. However, there are cases, where you indeed can use different addresses from the same server (if set up properly), while at the same time, even forcing custom SMTP can lead to cases, where one simply uses the same credentials at the root setting as well as the custom one.
In doubt, one could even write an extension, which simply adds its own nodemailer to force this functionality (basically what I am running at the moment).

My proposal would be to offer the option to override those settings (transport credentials as well as FROM address), but in a way, where it only helps people to use this within extensions - this way, it is more or less only available to rather experienced folks, while making sure people do not need to build very custom workaround-extensions, which is even more dangerous.
You can have a look at my implementation at https://github.com/jekuer/directus-discussion-11042

**Reply by christianrr** [reply]
@rijkvanzanten To be honest, when looking at it from a system-level perspective, I see things a bit differently. I also share @jekuer’s view that managing valid FROM addresses should ideally be the responsibility of the system operator. They should have the authority to define which sender addresses are valid for a specific Directus/SMTP setup.

In our case, for example, we would like to send emails from addresses like marketing@example.com, no-reply@example.com, or confirmation@example.com from within one Directus instance. This is a perfectly valid use case for any mail server that is correctly configured for a domain — including proper SPF and DMARC records. In fact, even sending from addresses with different domains can work without SPAM issues, as long as the relevant SPF records are in place.

I trust that system administrators are aware of the necessary requirements and capable of configuring them correctly. However, to prevent the issue you mentioned, I believe it's important to ensure that non-admin users should never be able to override the FROM address, e.g. in Flows. This capability should be restricted to custom extensions or environment configuration only.

The approach I proposed above, which involves exposing from in the `MailService`, would achieve exactly that: giving system operators more flexibility in custom extensions, while keeping the functionality securely out of reach for regular users.

**Reply by Kinzi** [reply]
I second @christianrr's use case.

As a "quick fix" just allowing to pass a name would already go a long way. something like "Rachel from company <email>" instead of just using the name of the instance.

In addition I have some whitelabel issues where it would be really cool if we could inject a different mail service in custom extensions.

**Reply by rijkvanzanten** [reply]
Making it available for custom extensions only is something we could do, as it already requires system-level admin access to install the extensions, at which point you have enough clearance. For regular in-app admins I can't expose it, as there's no guarantee the sys-admin has the smtp settings matching the set config in the mail service

**Reply by jekuer** [reply]
So, may I push my PR then? 😉
Happy to adjust it then, if any feedback occurs.

**Reply by jekuer** [reply]
I guess this is pretty much dead, right?
Just making sure whether I can trash my prepared PR #cleanup), @rijkvanzanten

**Reply by rijkvanzanten** [reply]
Oh my b! Left you hanging on read. More than welcome to open the PR 🙂

**Reply by jekuer** [reply]
Alright. Here we go: #25459
Happy to change or even trash things there. No hard feelings - just a proposal 😄 .

**Reply by Pachat** [reply]
> Making it available for custom extensions only is something we could do

We should also account for flows' Mail Header operations in this decision.
I'm using [email service provider] to send emails where I have a main address and multiple tenants' addresses, all properly verified. 

The issue I'm encountering is that Directus overrides the Sender object configuration.
Consequently, all emails are sent from the same email address without differentiating which tenant initiated the send operation.

Since I have control over the sender object within the flow configuration, I'm wondering about the rationale behind Directus overriding this setting?
I understand this might serve as a security measure for cloud/hosted versions to prevent abuse. 
However, for self-hosted instances, administrators should have the ability to control the sending address configuration.

**Reply by jekuer** [reply]
Hi @Pachat ,

The next release will include the option to provide your own FROM address.
It won't be available within the existing flow operation (as discussed)) and it will also only be about the address, not the underlying transporter (aka email server).

You can use this by building your own operation extension, where you import the mailService with:
```
const { MailService } = services;
const mailService = new MailService({ schema, accountability });
```

... and  the send the email with something like this:

```
mailService.send({ to, from, subject, html }).catch((error) => {
	throw new Error('Could not send mail in "mail" operation: ' + error);
});
```
The team put more advanced options on the roadmap, where you (hopefully) will be able to also define different sender profiles in the future - similar to how you can already define multiple different storages at the moment.

This adjustment will not resolve all issues, but at least is a first step towards more flexibility here. 😄

**Reply by inovosel** [reply]
So, even after the newest changes we don't have a simple option to fill in the "From" field in the Send Email flow operation?

**Reply by marksagangms-debug** [reply]
Any news about this?

**Reply by jekuer** [reply]
same as on Aug 4

**Comment by filipesmedeiros** [comment]
For anyone looking at this, setting up a hook and changing the `options.from` fields _does work_ it the most recent version. I set it up with the `{ name: string, address: string }` format.
